File: get_dividends_blog.py
import json

import requests
from base64 import b64encode

# Reutiliza a mesma sessão HTTP para manter cookies e evitar bloqueio por rate-limit
session = requests.Session()


# Função para buscar o trading name
def get_trading_name(ticker: str) -> str:
    # cria os parâmetros
    params: dict[str, str | int] = {
        "language": "pt-br",
        "pageNumber": 1,
        "pageSize": 20,
        "company": ticker,
    }

    # codifica os parâmetros como JSON e depois em base64
    string: str = b64encode(json.dumps(params).encode()).decode()

    # faz a requisição com os parâmetros
    r: requests.Response = session.get(
        r"https://sistemaswebb3-listados.b3.com.br/listedCompaniesProxy/CompanyCall/GetInitialCompanies/"
        + string,
    )

    # retorna o trading name da empresa, (é necessário remover pontos e barras)
    for i in r.json()["results"]:
        if i["issuingCompany"].lower() == ticker.lower():
            return i["tradingName"].replace("/", "").replace(".", "")

    # Se a empresa não for encontrada, retorna a mensagem de erro
    raise ValueError("Empresa não encontrada")


# Função para obter os proventos
def proventos(ticker: str) -> list[dict[str, str]]:
    trading_name: str = get_trading_name(ticker)
    results: list[dict[str, str]] = []
    n: int = 1
    while True:
        # cria os parâmetros com o trading name
        params: dict[str, str | int] = {
            "language": "pt-br",
            "pageNumber": n,
            "pageSize": 20,
            "tradingName": trading_name,
        }

        # codifica os parâmetros como JSON e depois em base64
        string: str = b64encode(json.dumps(params).encode()).decode()

        r: requests.Response = session.get(
            "https://sistemaswebb3-listados.b3.com.br/listedCompaniesProxy/CompanyCall/GetListedCashDividends/"
            + string,
        )

        if not r.text:
            break

        data: dict[str, list[dict[str, str]]] = r.json()
        if not data["results"]:
            break

        results += data["results"]
        n += 1
    return results


print(proventos("petr"))

File: get_dividends_blog.ts
// Interfaces para as respostas da API da B3

interface B3Page {
  pageNumber: number;
  pageSize: number;
  totalRecords: number | null;
  totalPages: number | null;
}

interface B3Company {
  codeCVM: string;
  issuingCompany: string;
  companyName: string;
  tradingName: string;
  cnpj: string;
  marketIndicator: string;
  typeBDR: string;
  dateListing: string;
  status: string;
  segment: string;
  segmentEng: string;
  type: string;
  market: string;
}

interface B3CompanyResponse {
  page: B3Page;
  results: B3Company[];
}

interface B3CashDividend {
  typeStock: string;
  dateApproval: string;
  valueCash: string;
  ratio: string;
  corporateAction: string;
  lastDatePriorEx: string;
  dateClosingPricePriorExDate: string;
  closingPricePriorExDate: string;
  quotedPerShares: string;
  corporateActionPrice: string;
  lastDateTimePriorEx: string;
}

interface B3DividendResponse {
  page: B3Page;
  results: B3CashDividend[];
}

// Params para as requisições

interface CompanySearchParams {
  language: string;
  pageNumber: number;
  pageSize: number;
  company: string;
}

interface DividendSearchParams {
  language: string;
  pageNumber: number;
  pageSize: number;
  tradingName: string;
}

// URLs base da API da B3
const B3_COMPANIES_URL =
  "https://sistemaswebb3-listados.b3.com.br/listedCompaniesProxy/CompanyCall/GetInitialCompanies/";
const B3_DIVIDENDS_URL =
  "https://sistemaswebb3-listados.b3.com.br/listedCompaniesProxy/CompanyCall/GetListedCashDividends/";

// Codifica os parâmetros como JSON e depois em base64
function encodeParams(params: CompanySearchParams | DividendSearchParams): string {
  return btoa(JSON.stringify(params));
}

// Função para buscar o trading name
async function getTradingName(ticker: string): Promise<string> {
  const params: CompanySearchParams = {
    language: "pt-br",
    pageNumber: 1,
    pageSize: 20,
    company: ticker,
  };

  const response = await fetch(B3_COMPANIES_URL + encodeParams(params));
  const data: B3CompanyResponse = await response.json();

  // Retorna o trading name da empresa (é necessário remover pontos e barras)
  const match = data.results.find(
    (company) => company.issuingCompany.toLowerCase() === ticker.toLowerCase()
  );

  if (!match) {
    throw new Error("Empresa não encontrada");
  }

  return match.tradingName.replaceAll("/", "").replaceAll(".", "");
}

// Função para obter os proventos
async function proventos(ticker: string): Promise<B3CashDividend[]> {
  const tradingName = await getTradingName(ticker);
  const results: B3CashDividend[] = [];
  let page = 1;

  while (true) {
    const params: DividendSearchParams = {
      language: "pt-br",
      pageNumber: page,
      pageSize: 20,
      tradingName,
    };

    const response = await fetch(B3_DIVIDENDS_URL + encodeParams(params));
    const text = await response.text();

    if (!text) break;

    const data: B3DividendResponse = JSON.parse(text);

    if (data.results.length === 0) break;

    results.push(...data.results);
    page++;
  }

  return results;
}

// Main
async function main(): Promise<void> {
  const dividends = await proventos("petr");
  console.log(dividends);
}

main();
